misk/src/main/kotlin/misk/web/jetty/ConnectionListener.kt

```diff
@@ -13,94 +12,140 @@ internal class ConnectionListener(
-  // with the high number of jetty threads. Instead we queue up changes to the connection and
-  // process the changes when metrics are refreshed by a single background thread.
   override fun onOpened(connection: Connection) {
-    openedConnections.add(ConnectionKey(connection))
+    activeConnections.add(ConnectionKey(connection))
+    metrics.activeConnections.labels(*labels).inc()
+    metrics.acceptedConnections.labels(*labels).inc()
   }
 
   override fun onClosed(connection: Connection) {
-    closedConnections.add(ConnectionKey(connection))
+    if (activeConnections.remove(ConnectionKey(connection))) {
+      // save the stats for the removed connection so they are included in the next call to
+      // refreshMetrics.
+      removedTotals.addConnection(connection)
+      recordDuration(connection)
+      metrics.activeConnections.labels(*labels).dec()
+    }
   }
 
   fun refreshMetrics() {
-    while (openedConnections.isNotEmpty()) {
-      connections.add(openedConnections.remove())
-      metrics.activeConnections.labels(*labels).inc()
-      metrics.acceptedConnections.labels(*labels).inc()
-    }
-
-    var bytesReceivedSnapshot = 0L
-    var bytesSentSnapshot = 0L
-    var messagesReceivedSnapshot = 0L
-    var messagesSentSnapshot = 0L
-
-    connections.forEach {
-      val bytesIn = it.connection.bytesIn
-      if (bytesIn > 0) bytesReceivedSnapshot += bytesIn
-
-      val bytesOut = it.connection.bytesOut
-      if (bytesOut > 0) bytesSentSnapshot += bytesOut
-
-      val messagesIn = it.connection.messagesIn
-      if (messagesIn > 0) messagesReceivedSnapshot += messagesIn
-
-      val messagesOut = it.connection.messagesOut
-      if (messagesOut > 0) messagesSentSnapshot += messagesOut
-
-      val connectionDuration = System.currentTimeMillis() - it.connection.createdTimeStamp
-      metrics.connectionDurations.record(connectionDuration.toDouble(), *labels)
+    val newTotals = Accumulator()
+    activeConnections.forEach {
+      newTotals.addConnection(it.connection)
+      recordDuration(it.connection)
     }
+    // add any connections that were removed since the previous refresh
+    newTotals.addSnapshot(removedTotals.snapshotAndReset())
 
     // Compute any diffs with the last time we took a snapshot, and apply those diffs
     // to the counter
-    val bytesReceivedDiff =
-        bytesReceivedSnapshot - totalBytesReceived.getAndSet(bytesReceivedSnapshot)
-    val bytesSentDiff =
-        bytesSentSnapshot - totalBytesSent.getAndSet(bytesSentSnapshot)
-    val messagesReceivedDiff =
-        messagesReceivedSnapshot - totalMessagesReceived.getAndSet(messagesReceivedSnapshot)
-    val messagesSentDiff =
-        messagesSentSnapshot - totalMessagesSent.getAndSet(messagesSentSnapshot)
-
-    if (bytesSentDiff > 0) {
-      metrics.bytesSent.labels(*labels).inc(bytesSentDiff.toDouble())
-    }
+    val newSnapshot = newTotals.snapshotAndReset()
+    val diff = newSnapshot - previousSnapshot
```

Comment:
I don’t know why everyone from 2004 hates operator overloading; this is nice!

Reply:
lol eh I'm not sold on it. more wanted to experiment with it.

is this really better than `newSnapshot.minus(previousSnapshot)`?

Reply:
I prefer it.